### DISCUSSION TITLE
How do I mock or override classes and methods for unittests?

### DISCUSSION THREAD

**Discussion by Subaku** [fastapi/fastapi#7920] [Questions]
I'm migrating my small API from Flask to FastAPI and one thing that's stopped working is my tests where I needed to mock and entire class a view would use to perform data operations. Now my mocking seems to just not work any longer. I see the docs mention something about overriding dependencies for tests and it makes it sound like I "should" be able to but I don't know how I might use it for my purpose.

Are there any examples of this?

To be clear I'm using Python's unittest and built in mock library. I'm trying to get back canned data from a call to a view via the TestClient. 

Thanks!

**Comment by Christian-Schultz** [comment]
This is a very guide using pytest: https://testdriven.io/blog/fastapi-crud/. Here pytest monkeypatch is used to mock behavior. It’s not Unittest, but if it’s not too much work you might want to change to pytest in any case.

**Comment by Subaku** [comment]
Thanks for the suggestion @Christian-Schultz. I'm trying out Pytest and it seems pretty great. Fixtures seem pretty awesome. I do like how easy it is to monkeypatch stuff but it doesn't allow for easy assertions around the things that need to be patched out.

I managed to get my tests running again by installing the great package pytest-mock. I also realized the reason for my mock failing was a case of what gets initialized when so I made it lazy and that fixed things. Still going to use Pytest though so thanks!

**Comment by tiangolo** [comment]
Thanks for the help here @Christian-Schultz ! :cake: :bowing_man: 

Thanks @Subaku for reporting back and closing the issue :+1:

**Comment by kricka91** [comment]
@Subaku Do you mind expanding on how you solved your problem? I'm also having trouble getting my mocks to run when migrating from python-flask to FastAPI.

**Comment by Subaku** [comment]
@kricka91 I installed the package pytest-mock and used a fixture to instantiate the mock I needed for my tests. So something like this:

``` python
class TestStuff:
    mock = None

    @pytest.fixture(autouse=True)
    def setup_tests(self, mocker):
        self.mock = mocker.patch(...).return_value

    def test_something(self):
        # Use self.mock
```

**Comment by kricka91** [comment]
@Subaku Thanks a lot for getting back so quickly!

I'm also using pytest, pytest-mock and fixtures. However, I'm applying the fixtures directly to each test function as different functions need to be mocked in different ways. I suspect my FastAPI app variable is loading its API endpoints before the mock is being applied. In particular, I'm curious about your statement "_I also realized the reason for my mock failing was a case of what gets initialized when so I made it lazy and that fixed things_"; what exactly did you change?

**Comment by Subaku** [comment]
You bet.

My comment about the "lazy" thing was how I had designed my views. I was needing to mock an adapter that my views would instantiate right away when FastAPI kicked on and gathered all the endpoints etc. Problem was I needed to mock said adapter and it was already instantiated so I couldn't mock it. I made it lazy via a property on the views so it would wait until it was first needed. I.e. after my test request was sent.

Afraid I've not dealt with your particular scenario before but one thing to maybe look at is the scoping feature of pytest. You can build fixtures that are scoped down to individual tests (I think...) so maybe that'll work. Sorry I couldn't be more help but best of luck to ya.
